Implement functions for following HTTP redirects

Avaq opened the pull request:
This feature will be implemented over ~two~ several commits. The initial commits support the last one. See the commit message bodies for extended explanations.

Avaq committed:
Rework the HTTP utilities for improved information preservation

Specifically, the 'request' function, which has been renamed to
'sendRequest', has two significant changes:

- It now returns the input request options along with the
  resulting IncomingMessage.
- The request body input now has to be wrapped in a Future to facilitate
  the former change: In order to return pure request options, the body
  must not yet have been initialized.

The reason for this rework is to faciliate the creation of new functions
that require *both* the Request and the IncomingMessage to perform their
tasks. Specifically, I am thinking of following redirects, where the
original request has to be re-sent with an altered URL.

To facilitate these changes, this commit introduces two new types:

- Request: The wrapper object for the request options.
- Response: The wrapper object for the Request/IncomingMessage pair.

Some functions have been reworked to operate on the Response type
instead of the IncomingMessage type.

Although this is a breaking change, for most uses of the library,
nothing will need to change in user land. This is because the interface
remained the same for most of the common use-cases, even though the data
types that flow through have been altered.

Avaq commented:
I'm quite proud of this approach. Here's the rendered README: https://github.com/fluture-js/fluture-node/blob/f4905343e858d01af2f6539c74aa3f12fbec09be/README.md

Avaq commented on index.js:
```diff
@@ -372,60 +433,300 @@ export const sendForm = method => url => headers => form => {
+  );
+};
+
+// TODO: Expand to include all headers that could cause servers to respond 304
+const conditionHeaders = ['etag'];
```
Besides this small detail which I need to look up, this code is ready.

dicearr commented on index.js:
```diff
@@ -96,7 +97,7 @@ export const streamOf = encase (buf => new Readable ({
 //.
 //. A [Readable][] stream which ends after emiting zero bytes. Can be useful
-//. as an empty [`request`](#request) body, for example.
+//. as an empty [`sendRequest`](#sendRequest) body, for example.
```
Shouldn't it be ``` [`request`](#request) ```?

Avaq replied:
No, but it might have to be ``` [`Request`](#Request) ```?

dicearr commented on index.js:
```diff
@@ -372,60 +433,300 @@ export const sendForm = method => url => headers => form => {
+//. const myFollowRedirects = followRedirectsWith ({
+//.   300: myMultipleChoiceRedirectionStrategy,
+//.   301: redirectOriginalRequest,
+//.   400: tryAgainAfterSomeTime,
```
I like the idea of not imposing any particular meaning to the status codes. However, in this specific example, I find conceptually confusing the mix of retrying and redirecting strategies.

I get that, with the current approach, users are going to be able to implement this anyways. So it might be a good idea to give them an example.

Avaq replied:
Maybe I'll add a comment that it's impractical to mix redirection and retry logic in the same function, and that the code shown is just in indication of what's possible.

Avaq replied:
Actually, now I'm thinking of simplifying the types a bit and making the whole thing more focused on redirection. The retry and error handling stuff will no longer be possible.

davidchambers commented on index.js:
```diff
@@ -372,60 +433,300 @@ export const sendForm = method => url => headers => form => {
+export const retryWithoutCondition = response => {
+  const original = Response.request (response);
+  const options = Request.options (original);
+  const headers = Object.entries (options.headers || {});
```
I look forward to getting to use [`??`][1] one day. :)


[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator

Avaq replied:
Mee too!

davidchambers commented on index.js:
```diff
@@ -372,60 +433,300 @@ export const sendForm = method => url => headers => form => {
+//. [Http](#http) section shows this.
 export const acceptStatus = code => res => (
-  code === res.statusCode ? resolve (res) : reject (res)
+  code === Response.message (res).statusCode ? resolve (res) : reject (res)
```
`f (x)` formatting does not work well in cases such as this one. One option is to use redundant parens to make the grouping clear (e.g. `(f (x)).y`).

Avaq committed:
Add capability for additional endpoints to testing server

Avaq committed:
Add a function for pattern-matching on response status codes

The function, named 'matchStatus', is an excellent building block for
all sorts of use-cases, and will be used in the upcoming redirect
following feature.

Avaq committed:
Rework the HTTP usage example using matchStatus

This replaces the old approach which used acceptStatus.

It also expands the explanation to talk about each of the functions
used.

Avaq commented:
I'm taking this into a slightly different direction. But I think it's even more exciting! In the new approach, `followRedirectsWith` will be implemented on top of the new `matchStatus` function, which I have already pushed. I think this function just really puts the library together. It was the missing piece when it comes to simplicity meeting convenience. And it turns out it can serve lots of fantastic use-cases!

Avaq commented:
The currently implementation hangs on the fact that it's currently difficult to say whether one `Request` is equivalent to another. I have the solution thought out and will implement it soon.

Avaq committed:
Add mechanism for normalizing Request options

The underlying request API exposed by Node applies some transformations
to the options before using them, and also uses a mechanism for merging
the URL with the options:

https://nodejs.org/api/http.html#http_http_request_options_callback

In order to make this library more consistent, I've added the a
function which returns the request options such that conflicting values
are less common, and so that it's also less common to be able to
represent the same request in multiple ways.

Avaq committed:
Add functions for following redirects

Adds a framework for redirection of requests. Add the core of the idea
is one new function: 'followRedirectsWith'. This function takes what I'm
calling a "rediction policy" in the form of a lambda that takes a
Response and produces a (new) Request from it ('Response -> Request').

Using the redirection policy and a max number, the followRedirectsWith
recursively follows the new Requests produced, until either the maximum
number of redirects is reached, or the policy produces a request which
has already been seen in the chain of redirections.

This last principle means we are protected against infinite redirect
loops, but it also gives plicies a way to "not redirect" by simply
returning the original request.

There are two built-in redirection policies:

- The 'defaultRedirectionPolicy' policy follows RFC2616 strictly.
- The 'aggresiveRedirectionPolicy' policy deviates a little from RFC2616
  to better facilitate fully automated programs.

These policies are composed from smaller policies using the matchStatus
function, and users have access to the underlying 'Response -> Request'
function to build their own policies, if they wish.

Avaq commented:
This is PR ready using the new approach. Here's the [rendered README](https://github.com/fluture-js/fluture-node/blob/b82cede90f4d43c8e79f3dbab66f0710aeea2345/README.md).

Avaq commented on index.js:
```diff
@@ -188,45 +195,74 @@ export const immediate = x => Future ((rej, res) => {
+//. Note that we were in control of the following:
+//.
+//. - How redirects are followed: We use [`followRedirects`](#followRedirects)
+//.   with a maxmum of 20 redirects, but we could have used a custom
```
```suggestion
//.   with a maxmum of 20 redirects, but we could have used a different
```

Avaq commented on index.js:
```diff
@@ -372,66 +490,338 @@ export const sendForm = method => url => headers => form => {
+//. original request used the GET method, and redirects with status code 304
+//. are left alone for a caching layer to deal with.
+//.
+//. See also [`aggresiveRedirectionPolicy`](#aggresiveRedirectionPolicy).
+//.
+//. This redirection policy is used by default in the
+//. [`followRedirects`](#followRedirects) function. You can extend it, using
+//. [`matchStatus`](#matchStatus) to create a custom redirection policy, as
+//. show in the example:
```
Swap these paragraphs around.

Avaq replied:
```suggestion
//. shown in the example:
```

Avaq commented on index.js:
```diff
@@ -372,66 +490,338 @@ export const sendForm = method => url => headers => form => {
+//. first argument.
+//.
+//. The resulting function `Response -> a` has the same signature as the input
+//. functions, meaning you can use `matchStatus` *again* to "extend" the a
```
```suggestion
//. functions, meaning you can use `matchStatus` *again* to "extend" the
```

Avaq reviewed (commented):
Some README improvements I want to make

davidchambers commented:
The example is extremely compelling! Fantastic work, Aldwin!

Avaq commented on index.js:
```diff
@@ -372,66 +490,338 @@ export const sendForm = method => url => headers => form => {
+  307: redirectIfGetMethod,
+});
+
+//# aggresiveRedirectionPolicy :: Response -> Request
```
Aggressive is spelled with two s's.

Avaq commented:
I reviewed this with some colleagues at work.
